### DISCUSSION TITLE
Easiest method to install yazi into an arm64 debian (raspberry pi)

### DISCUSSION THREAD

**Discussion by fcaraujo** [sxyazi/yazi#1409] [1. Q&A]
Hey everyone,

I'm running debian (arm64) on a rpi and and was not able to install yazi in it.
I searched and found no discussion about it.

I tried through snap, downloading the file and running install, got the following error:
```
error: cannot install snap file: snap "yazi" supported architectures (amd64) are incompatible with
       this system (arm64)
```  

Never contributed/managed any package available to be installed through apt (that would be my first guess);
Also, I've never developed in Rust, but I'd guess I could try to build/publish it there. (?)
Thanks

**Comment by sxyazi** [comment]
You can try building Yazi by following the instructions at https://yazi-rs.github.io/docs/installation#build-from-source.

As for Snap, its code is available at https://github.com/sxyazi/yazi/blob/main/snap/snapcraft.yaml, but I've never used Snap myself, so I'm not sure how to make it support ARM64. 

cc @gapplef, do you have any pointers?

**Comment by gapplef** [comment]
Arm64 support of snap was disabled as I don't have any arm device with Linux, and can't do the test myself.
@fcaraujo could you test the snap-arm64 from https://github.com/gapplef/yazi/actions/runs/10245375735

**Comment by gapplef** [comment]
@sxyazi I would first update the snap to core24 https://github.com/sxyazi/yazi/pull/1415

**Comment by fcaraujo** [comment]
Thanks @sxyazi and @gapplef.

After getting the rust tool belt (and already feeling the "cool kid" mood of doing sth new) I tried to build it once.
However, my RPi became unresponsive, and I was not able to ssh back into it and check whether it succeeded or not.

I'm retrying it at the moment, I noticed at the beginning it was quite CPU-intensive:
![image](https://github.com/user-attachments/assets/54585fb6-065d-447a-8ca7-e0e96d416e43)

I wonder if that actually froze my instance, although I'm leaving it unattended for a while again and coming back later 🤞; tomorrow, I might have good or bad news.

@gapplef sure thing I'm happy to help! 
I just had a quick glance at the logs to see if I'd figure out which commands I need to reproduce here; 
I'll just need some guidance since I've never used snap myself as well.
Could you please share some baby-step instructions?

**Reply by gapplef** [reply]
Just get the snap package from https://github.com/gapplef/yazi/actions/runs/10245375735 
and `sudo snap install --dangerous --classic <package_name.snap>`

**Comment by fcaraujo** [comment]
Good news, after a couple of attempts my RPi survived and built yazi (after ~45 min running it 🫠).
Apparently it's working like a charm!

Just out of curiosity, is it normal taking this long for other platforms as well (I suppose hardware dependency)?
Is there anything cli-related I could pass/configure to speed things up in a personalised way for raspberrypi?

RE testing: as my instance's already in a working state and I wanna isolate things, I'll spin up an empty environment here and test your snap package soon.
Basically, I just need `snapd` installed, download the snap from it and execute it, right?
@gapplef I see the artifact has expired - is there a way I can re-run it, or can you kick them off again, pls?

**Reply by gapplef** [reply]
Yes, you are right, and you may find helps here https://snapcraft.io/docs/installing-snap-on-debian
I have re-run the build, and the artifacts will expired in one day, so you may download it first and do the test later.
https://github.com/gapplef/yazi/actions/runs/10279150280

**Comment by fcaraujo** [comment]
@gapplef can you pls check again the artifacts? I downloaded the arm64 one (zipped) and when extracted i found an amd64.snap file; I tried regardless its name to confirm what I suspected:
 
```
fc@yazi:~/Downloads $ sudo snap install --dangerous --classic yazi_nightly_amd64.snap
error: cannot install snap file: snap "yazi" supported architectures (amd64) are incompatible with
       this system (arm64)
```

**Comment by fcaraujo** [comment]
BTW @gapplef I had a quick look at the arm64 job raw logs now, and it really packed an *amd64.snap file (I just search for "and"); so at least I could check I'm not crazy or screwed sth there heheheh.
 
I don't understand how it works and I couldn't spot anything arch/platform-related to determine whether it's missing some parameter/configs though... : (

My guess would be some copy/paste from the amd job that's missing or the runner's architecture mismatching the target or whatnot?!

**Comment by fcaraujo** [comment]
Also, I'm not very familiar with these workflow/actions, but I'll try to have some read next (and then I won't be making these wild guesses anymore... 😅) 

However, I noticed you used both archs in the workflow file, (that had 2 diff jobs tho) but the path passed was `path: yazi*.snap`.

I'm not sure if in case there are multiple files (ending with diff suffixes) that would upload them all, or just pick up the first one (and then alphabetically `amd` would come first).

And I had a glance at snapcraft's docs, noticed there are some cases that it builds both for amd and arm as well at the same time.

It might be useful for troubleshooting reasons, if we could make it a bit more verbose and show what's doing under the hood.

And if it's possible I'd suggest we increase the retention to 2d, so if we have someone looking into it and misses it 1d, it would still be there without bothering you to re-run anything.

Anyway I'll let you troubleshoot and I hope my assumptions could get closer to a resolution.
Anything just let me know if I can help @gapplef .

**Comment by gapplef** [comment]
Basically, I specific platforms to `amd64` and `arm64` in `snap/snapcraft.yml`,
```yaml
platforms:
  amd64:
  arm64:
```
and use both arch in the workflow. 

I excepted it to build for amd64 on amd64, and build for arm64 on arm64, automatically. But seems it always build amd64 on arm64.....
For reference you can see the docs of build plans: https://forum.snapcraft.io/t/explanation-architectures/36759


As for multiple files, I think it will upload all of them, as in the case of this run
https://github.com/gapplef/yazi/actions/runs/10297421494

I'm still trying to figure out the reason, but you may first test the file in the above link.

**Comment by fcaraujo** [comment]
Righto I already downloaded both, and I'll give it a go tomorrow.
Btw I noticed it's double the size, so I suppose both zips have both amd/arm files in them.
I'll keep you posted as I go. O/

**Comment by gapplef** [comment]
Yeah, I think two zips have the same files.

I am not sure why, but seems it can't build arm64 snap on arm64. I only managed to build arm64 on amd64.

And, since I bundle necessary dependencies(`fd-find`, `ripgrep`, `zoxide`, `fzf`) with the snap, I'm not sure whether those binaries are `arm64` version when build arm64 snap on amd64 system 😂 You can test it out.

**Comment by fcaraujo** [comment]
Oh right, gotcha! 😅

I have a hunch that trying to build in a diff architecture would get out of whack and not be able to execute, but we will figure this out.

For now, I was at least able to install it:

```
fc@yazi:~/Downloads $ cat /etc/os-release
PRETTY_NAME="Debian GNU/Linux 12 (bookworm)"
NAME="Debian GNU/Linux"
VERSION_ID="12"
VERSION="12 (bookworm)"
VERSION_CODENAME=bookworm
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"

fc@yazi:~/Downloads $ uname -a
Linux yazi 6.6.31+rpt-rpi-v8 #1 SMP PREEMPT Debian 1:6.6.31-1+rpt1 (2024-05-29) aarch64 GNU/Linux

fc@yazi:~/Downloads $ sudo snap install --dangerous --classic yazi_nightly_arm64.snap
2024-08-09T10:08:28+12:00 INFO Waiting for automatic snapd restart...
yazi nightly installed

fc@yazi:~/Downloads $ yazi -V
cannot snap-exec: cannot exec "/snap/yazi/x1/yazi": exec format error

fc@yazi:~/Downloads $ which yazi
/snap/bin/yazi

fc@yazi:~/Downloads $ yazi
cannot snap-exec: cannot exec "/snap/yazi/x1/yazi": exec format error
```

BTW, from memory, I saw this `exec format error` before when the bin we're trying to execute is in the wrong arch.

**Comment by gapplef** [comment]
I can't figure it out. Let's get some help 🤣
https://forum.snapcraft.io/t/help-needed-cant-build-arm64-snap-with-github-actions/41356

**Comment by jnsgruk** [comment]
Just seen this after I posted a response on the Discourse: https://forum.snapcraft.io/t/help-needed-cant-build-arm64-snap-with-github-actions/41356/3

Let me know if I can help here. Would love to see the snap publicly available.

**Reply by jnsgruk** [reply]
> Just seen this after I posted a response on the Discourse: https://forum.snapcraft.io/t/help-needed-cant-build-arm64-snap-with-github-actions/41356/3
> 
> Let me know if I can help here. Would love to see the snap publicly available.

Related: https://github.com/sxyazi/yazi/pull/2188

**Comment by sxyazi** [comment]
Yazi now supports snap packages for arm64: https://github.com/sxyazi/yazi/releases/tag/nightly

Thanks @jnsgruk 🎉🎉
